### ISSUE TITLE
[css-sizing-4] Clarify spec about when elements are allowed to not have a last remembered size

### ISSUE THREAD

**Issue by dholbert** [w3c/csswg-drafts#6220]
Currently [css-sizing-4 section 5.2.1 "Last Remembered Size"](https://drafts.csswg.org/css-sizing-4/#last-remembered) says:
> An element might not have a last remembered size, if it has never been rendered without size containment. (In this case, it will instead use the fallback value provided along with auto.)

This statement logically implies that its contrapositive must hold, i.e. it seems to be saying: "If an element **has ever been rendered without size containment**, then it **must have a last remembered size**".

I'm not sure that we want to require this, though. Consider e.g. these scenarios:

(1) an element is rendered, and then becomes `display:none` (losing its css boxes).  And then it gains `contain:size`, and then loses `display:none` (e.g. going back to `display:block`).  Do we want to require (or imply) that its last-rendered-size should survive this process?

(2) an element is rendered (gaining a Last Remembered Size), and then gains `contain:size`, and then has its `display` value changed in a slightly-less-destructive way (e.g. from `display:inline` to `display:grid`).  Do we want to require/imply that its last-rendered-size should survive this process?

(3) an element is rendered, and then gains `contain:size`, and then is reparented to somewhere completely different in the DOM tree.  Do we want to require (or imply) that this element's last remembered size must survive this process?

In all three of these cases, the element *has* previously been rendered without size containment, so the spec's caveat about "an element might not have a last rendered size..." seems not to apply, i.e. it seems like the spec assumes that the element *would* have a last rendered size. But I'm not sure if that's intentional or not.

(Put another way: is the intent here that the last-rendered-size is something attached to the node, or something attached to the CSS box?)

**Comment by cbiesinger** [comment]
FWIW, in terms of Blink implementation, we have decided that we have to use an actual resize observer and store the last remembered size in the DOM node. We have not thought of a way to get the right semantics without actually using resize observer (container queries make it hard).

(cc @bfgeek @vmpstr @tabatkins )

**Comment by Loirooriol** [comment]
Also, what about a dynamic `contain-intrinsic-size`? Like

1. You render an element  with `contain-intrinsic-size: auto 0px` and no containment, so it remembers the size, e.g. 100px.
2. You set `contain-intrinsic-size: none`.
3. You change the contents so that the inner size becomes e.g. 200px. This is not recorded as the last remembered size.
4. You add `contain: size; contain-intrinsic-size: auto 0px`.

Do the 100px survive this process?

> we have to use an actual resize observer and store the last remembered size in the DOM node

How do you handle ::before and ::after? Can you add a resize observer to a pseudo-element? AFAIK Blink doesn't create pseudo-element nodes if they are not going to generate a box, and disposes of the nodes if they stop generating a box...

**Comment by chrishtr** [comment]
> Do the 100px survive this process? (@Loirooriol)

No. Once you remove the `auto` keyword it loses memory of the past. Do you see an issue with this behavior?

> (Put another way: is the intent here that the last-rendered-size is something attached to the node, or something attached to the CSS box?) @dholbert 

I agree with what @cbiesinger said: it should be a property of the element, not the box. Otherwise its behavior will I think be too unpredictable for developers.

**Comment by chrishtr** [comment]
> How do you handle ::before and ::after? Can you add a resize observer to a pseudo-element? AFAIK Blink doesn't create pseudo-element nodes if they are not going to generate a box, and disposes of the nodes if they stop generating a box...

No I don't think you can attach a resize observer to a pseudo-element. Therefore I think we should amend the spec to exclude pseudo-elements from the auto behavior.

**Comment by Loirooriol** [comment]
I don't see a problem with clearing the remembered size when removing the `auto` keyword, that's what I would expect. But the spec should say so.

**Comment by chrishtr** [comment]
Ok. Proposed spec edits:
* `auto` behavior only applies to elements which can be `ResizeObserver` targets (in particular, excluding pseudo elements for now)
* `auto` value is forgotten when another a `ResizeObsever` opportunity occurs again and `auto` is not present at that time (meaning the update-the-rendering steps occurred).

@tabatkins WDYT?

**Comment by Loirooriol** [comment]
No strong opinion, but maybe I would even expect to remove the last remembered size during the first layout after removing `auto`, even if there hasn't been any `ResizeObserver` opportunity.

```js
// Assume last remembered size is 100px
element.style.containIntrinsicSize = "none";
element.offsetLeft; // Forces sync layout, clears remembered size?
element.style.containIntrinsicSize = "auto 0px"; // Re-add "auto" before ResizeObserver opportunity
element.style.contain = "size";
```

**Comment by tabatkins** [comment]
Yeah, given that we're already leaning heavily on RO timing for this, if possible I'd like to stick with that; it keeps things simple and predictable.

I've gone ahead and editted in the conclusions summarized by @chrishtr in <https://github.com/w3c/csswg-drafts/commit/22a5e0e435a74e6b1ef7db2d2b72e6d405460d1a>. Lmk if that satisfies everyone!

**Comment by chrishtr** [comment]
@tabatkins LGTM!

**Comment by cbiesinger** [comment]
@dholbert could you confirm that the latest version satisfies your concerns>?

**Comment by dholbert** [comment]
Yeah, this seems like a reasonable way to address this.

Minor nitpick: right now, this issue's added text (in https://drafts.csswg.org/css-sizing-4/#last-remembered ) is worded as follows:
> At the time that ResizeObserver events are determined and delivered [...] record the current inner dimensions of its principal box as its last remembered size.

There's a little ambiguity here, because if there are no ResizeObservers, then arguably there's not **any** time when ResizeObserver events are determined and delivered.  (There are no events to be determined or delivered.)  So that sort of implies that last-remembered-sizes just never get recorded in a document without ResizeObservers.

Maybe this should instead be phrased more hypothetically, like so:
> At the time that ResizeObserver events would be determined and delivered (if there were ResizeObservers registered): [...]

But maybe not a big deal. *shrug*

**Comment by Loirooriol** [comment]
In #7527 we resolved

> If, when an element goes from not generating a principal box to generating one, it does not have cis:auto, forget any last remembered size

Given that, I'm now more convinced that the last remembered size should be forgotten during the first layout after removing `auto`, even if there hasn't been any ResizeObserver opportunity, as I said in https://github.com/w3c/csswg-drafts/issues/6220#issuecomment-830711439.

Otherwise it doesn't look consistent:

```js
// element has last remembered size
element.style.containIntrinsicSize = "none";
element.offsetLeft;
// element still has last remembered size
```

```js
// element has last remembered size
element.style.display = "none";
element.offsetLeft;
element.style.display = "";
element.style.containIntrinsicSize = "none";
element.offsetLeft;
// element doesn't have a last remembered size
```

Also, it's simpler from an implementation point of view to just clear the last remembered size when laying out.

I don't buy this argument:

> The problem with this definition is that it's not clear how to define it. e.g. what if some other element's offsetLeft was queried - does that require layout on element or not?

<details>
<summary>Since unless I'm missing something, whether offsetLeft causes layout is already webexposed e.g. with transitions</summary>

```js
element.style.transition = "width 1s";
element.style.width = "100px";
element.style.width = "200px";
element.clientWidth; // 200 (if we started with width:auto)
```
```js
element.style.transition = "width 1s";
element.style.width = "100px";
element.offsetLeft;
element.style.width = "200px";
element.clientWidth; // 100
```

</details>

Agenda+ to discuss removing the last remembered size without waiting for ResizeObserver when laying out without c-i-s:auto.

**Comment by bfgeek** [comment]
(I might miss the meeting this morning - likely be around for the end - but want to make sure that we generally keep ResizeObserver timing for this feature). The ResizeObserver timing is important for a few reasons.
1. The lifetime of boxes aren't defined. E.g. all engines have subtle rules around when the box-tree is re-built, when nodes are re-used, etc. For example in Blink fieldsets can cause re-attachment where in other implementations they don't (one among many examples).
2. What APIs trigger layout (and how much layout) isn't defined. E.g. for offsetLeft/etc. you can currently have a conforming implementation which performs a partial layout just to answer the offsetLeft query for example. We shouldn't tie more side-effects to forced-sync-layouts lightly.

**Comment by Loirooriol** [comment]
Well, even if the lifetime of boxes and the APIs that trigger layout aren't defined, in #7527 we resolved

> If, when an element goes from not generating a principal box to generating one, it does not have cis:auto, forget any last remembered size

so we already depend on that.
